keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,870 @@
+underlying `Service`(s) across the supercluster.
+
+#### EndpointSlice
+
+When a `ServiceExport` is created, this will cause `EndpointSlice` objects for
+the underlying `Service` to be created in each cluster within the supercluster.
+One or more `EndpointSlice` resources will exist for each cluster that exported
+the `Service`, with each `EndpointSlice` containing only endpoints from its
+source cluster. These `EndpointSlice` objects will be marked as managed by the
+supercluster service controller, so that the endpoint slice controller doesn’t
+delete them.
+
+```
+<<[UNRESOLVED]>>
+We have not yet sorted out scalability impact here. We hope the upper bound for
+imported endpoints + in-cluster endpoints will be ~= the upper bound for
+in-cluster endpoints today, but this remains to be determined.
+<<[/UNRESOLVED]>>
+```
+
+#### Endpoint TTL
+
+To prevent stale endpoints from persisting in the event that a cluster becomes
+unreachable to the supercluster controller, each `EndpointSlice` is associated
+with a lease representing connectivity with its source cluster. The supercluster
+service controller is responsible for periodically renewing the lease so long as
+the connection with the source cluster is confirmed alive. A separate
+controller, that may run inside each cluster, is responsible for watching each
+lease and removing all remaining `EndpointSlices` associated with a cluster when
+that cluster’s lease expires.
+
+### Consumption of EndpointSlice
+
+To consume a supercluster service, users will use the domain name associated
+with their `ServiceExport`. When the mcsd-controller sees a `ServiceExport`, a
+`ImportedService` will be introduced, which can be largely ignored by the user. 
+
+An `ImportedService` is a service that may have endpoints in other clusters.
+This includes 3 scenarios:
+1. This service is running entirely in different cluster(s)
+1. This service has endpoints in other cluster(s) and in this cluster
+1. This service is running entirely in this cluster, but is exported to other cluster(s) as well
+
+For each exported service, one `ServiceExport` will exist in each cluster that
+runs the service. The mcsd-controller will create and maintain a derived
+`ImportedService` in each cluster within the supercluster (see: [constraints and
+conflict resolution](#constraints-and-conflict-resolution)). If all `ServiceExport` instances are deleted, each
+`ImportedService` will also be deleted from all clusters.
+
+Since a given `ImportedService` may be backed by multiple `EndpointSlices`, a
+given `EndpointSlice` will reference its `ImportedService` using the label
+`multicluster.kubernetes.io/imported-service-name` similarly to how an
+`EndpointSlice` is associated with its `Service` in a single cluster. Each
+imported `EndpointSlice` will also have a
+`multicluster.kubernetes.io/source-cluster` label with a registry-scoped unique
+identifier for the cluster.
```

Comment:
Do any existing core cluster-components such as kube-proxy/CoreDNS/kube-dns need to be modified to ignore these new `EndpointSlices`?
What about ingress-controllers, prometheus / kube-prometheus, and CNI related daemons?
Will using `EndpointSlices` in this way potentially affect user's scripts, automation, or integrations?

Have we considered also cloning the API into a CRD like `ImportedEndpointSlices` in the same way we're doing for `Services`?

This would make the UX for `kubectl get ep,importedendpoints` match `kubectl get svc,importedservice` instead of conflating one but not the other.

Reply:
Existing components should not require any modification to ignore these new slices - that was one driver for using `ImportedService` instead of trying to extend `Service`. One of the nice features of `EndpointSlice` compared to `Endpoints` is that they are not strictly coupled to a Service, so as long as we don't use the `kubernetes.io/service-name` label, our slices should have no impact on existing components that discover them by association with a `Service`.

Reply:
modulo scalabiity aspects (which in general we didn't yet solve), because things like dns, kube-proxy etc. watch for all endpoints (so may oom or sth...)

Reply:
Good point - though existing logic shouldn't be impacted by these new unlabelled slices (at least from the Service API's POV), they will still trigger events and could potentially have noticeable side effects at scale.